db/migrations/versions/027_start_date_before_end.py

```diff
@@ -0,0 +1,56 @@
+"""require start_dates to be before end_dates, if both fields are present in a row
+
+Revision ID: 027_start_date_before_end
+Revises: 026_standardise_actual_forecast
+Create Date: 2024-04-04 13:59:15.747254
+
+"""
+
+from alembic import op
+
+# revision identifiers, used by Alembic.
+revision = "027_start_date_before_end"
+down_revision = "026_standardise_actual_forecast"
+branch_labels = None
+depends_on = None
+
+
+def upgrade():
+    for table in ["funding", "programme_funding_management", "project_progress"]:
+        op.create_check_constraint(
+            "start_before_end",  # gets prefixed with `ck_{table}`
+            table,
+            "start_date IS NULL OR end_date IS NULL OR (start_date <= end_date)",
+            postgresql_not_valid=True,
+        )
+
+    for table in ["outcome_data", "output_data"]:
+        op.create_check_constraint(
+            "start_before_end",  # gets prefixed with `ck_{table}`
+            table,
+            "(start_date <= end_date)",
+            postgresql_not_valid=True,
+        )
+
+    for table in ["submission_dim"]:
+        op.create_check_constraint(
+            "start_before_end",  # gets prefixed with `ck_{table}`
+            table,
+            "(reporting_period_start <= reporting_period_end)",
+            postgresql_not_valid=True,
+        )
```

Comment:
Just a thought, but won't this blow-up if the constraint is not met somewhere in our existing data? What's therefore the need to have a separate validate script for checking the constraint?

Reply:
The `postgresql_not_valid=True` bits mean that the constraint goes in but ignores any data that already exists which isn't valid. The constraint is only validated on new data.

The second migration then validates the constraint and blows up on bad data. It being separate means:

* we keep the constraint which validates _new_ data
* each migration takes a less heavy-handed lock, keeping database queries less restricted during migration